frontend/src/metabase/query_builder/components/DataSelector/DataSelector.jsx

```diff
@@ -455,7 +455,12 @@ export class UnconnectedDataSelector extends Component {
       await this.switchToStep(DATABASE_STEP);
     } else if (this.state.selectedTableId && steps.includes(FIELD_STEP)) {
       await this.switchToStep(FIELD_STEP);
-    } else if (this.state.selectedSchemaId && steps.includes(TABLE_STEP)) {
+    } else if (
+      // Schema id is explicitly set when going through the New > Question/Model flow,
+      // whereas we have to obtain it from the state when opening a saved question.
+      (this.state.selectedSchemaId || this.state.selectedSchema?.id) &&
```

Comment:
This line here is the actual fix.
Everything else is related to tests.

Reply:
@ranquild no matter what I do, there's always some part of the internal entities state that's not happy. This particular solution works perfectly for multi-schema databases. There is an E2E test that confirms that.

But there are some other E2E tests that fail because we're not fetching`/api/database/${SAMPLE_DB_ID}/schema/PUBLIC` anymore. This can also result in only the selected table being shown
![image](https://github.com/metabase/metabase/assets/31325167/b19caf27-f130-4504-ba54-7aff12385ed0)

If you have any other ideas how to approach this, please LMK.
Also, feel free to commit directly to the branch.

Reply:
As per our previous conversation, one of the attempts at fixing this problem was setting only the `selectedSchemaId` but in practice it was still often `undefined` so it didn't solve the multi-schema problem.
https://github.com/metabase/metabase/pull/39930/commits/d55b91596c0e94e6867d8a3900eeca0af5f9a91c#diff-b8d140ac6041c995aeefb03f0ac704c4ec7410ef7772858b93f4dec5a57fc7c6R265-R267

Reply:
> If you have any other ideas how to approach this, please LMK.
> Also, feel free to commit directly to the branch.

Btw, this is my attempt at solving the problem. LMK if the approach makes sense:
https://github.com/metabase/metabase/pull/39930/commits/83ce3c249a76502dbd4dc89a38f75c22a7030bcb